### ISSUE TITLE
Mongock and CosmosDB

### ISSUE THREAD

**Issue by leonelsanchesdasilva** [cloudyrock/mongock#256]
We are trying to deploy an application that uses Mongock to CosmosDB. According to Azure Docs, CosmosDB creates all the indexes automatically for the collections in a database: https://docs.microsoft.com/en-us/azure/cosmos-db/mongodb-indexing. We are also aware about the advise given by @dieppa on #181 about creating indexes.

We performed two steps: trying to run the application without any collections created, and trying to run the application with `mongockLock` collection created. 

On the first step, we had this error (no collections created, only the database):

```
: Sending command '{"createIndexes": "mongockLock", "indexes": [{"key": {"key": 1}, "name": "key_1", "ns": "sampleapp.mongockLock", "unique": true}]}' with request id 28 to database sampleapp on connection [connectionId{localValue:3, serverValue:44859925}] to server cdb-ms-prod-centralus1-fd15.documents.azure.com:10255
2020-10-07 20:06:09.869 DEBUG 7160 --- [           main] org.mongodb.driver.protocol.command      : Execution of command with request id 28 completed successfully in 5.69 ms on connection [connectionId{localValue:3, serverValue:44859925}] to server cdb-ms-prod-centralus1-fd15.documents.azure.com:10255
2020-10-07 20:06:09.869 DEBUG 7160 --- [           main] g.c.m.d.m.s.v.r.MongoSync4RepositoryBase : Index in collection [mongockLock] was recreated
2020-10-07 20:06:09.869  WARN 7160 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'mongockConnectionDriver' defined in class path resource [com/github/cloudyrock/spring/v5/MongockSpringDataV3CoreContext.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.github.cloudyrock.mongock.driver.mongodb.springdata.v3.SpringDataMongo3Driver]: Factory method 'mongockConnectionDriver' threw exception; nested exception is io.changock.migration.api.exception.ChangockException: Max tries 3 index  creation
2020-10-07 20:06:09.885  INFO 7160 --- [           main] org.mongodb.driver.connection            : Closed connection [connectionId{localValue:3, serverValue:44859925}] to cdb-ms-prod-centralus1-fd15.documents.azure.com:10255 because the pool has been closed.
2020-10-07 20:06:09.885 DEBUG 7160 --- [           main] org.mongodb.driver.connection            : Closing connection connectionId{localValue:3, serverValue:44859925}
2020-10-07 20:06:09.885 DEBUG 7160 --- [azure.com:10255] org.mongodb.driver.connection            : Closing connection connectionId{localValue:2, serverValue:259432963}
2020-10-07 20:06:09.978 DEBUG 7160 --- [isson-netty-2-2] io.netty.buffer.PoolThreadCache          : Freed 3 thread-local buffer(s) from thread: redisson-netty-2-2
```

On the second setp, we had this error (`mongockLock` collection created):

```
Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'mongockConnectionDriver' defined in class path resource [com/github/cloudyrock/spring/v5/MongockSpringDataV3CoreContext.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.github.cloudyrock.mongock.driver.mongodb.springdata.v3.SpringDataMongo3Driver]: Factory method 'mongockConnectionDriver' threw exception; nested exception is io.changock.migration.api.exception.ChangockException: Index creation not allowed, but not created or wrongly created for collection mongockLock
        at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:655) ~[spring-beans-5.2.8.RELEASE.jar!/:5.2.8.RELEASE]
```

Is there something else we can try to make Mongock work with CosmosDB?

**Comment by dieppa** [comment]
Hello @leonelsanchesdasilva ,

if I'm not wrong, the CosmosDB documentation says that it creates the `_id` index automatically, as MongoDB also does, but you are free to create any other index you wish. It won't do it as it doesn't know what to create.


Assuming you set `index-creation=false` and you are not creating the required indexes manually, it's totally  normal that Mongock fails. You either allow Mongock to take care of it, or do it yourself.

This makes me aware that in Mongock's documentation doesn't say which the indexes to be created are(my fault). I need to add that. But meanwhile I give them to you here. It should be similar to the following indexes.

#### For`mongockChangeLog` collection
```json
[
	{
		"v" : 2,
		"key" : {
			"_id" : 1
		},
		"name" : "_id_",
		"ns" : "my-database.mongockChangeLog"
	},
	{
		"v" : 2,
		"unique" : true,
		"key" : {
			"executionId" : 1,
			"author" : 1,
			"changeId" : 1
		},
		"name" : "executionId_1_author_1_changeId_1",
		"ns" : "my-database.mongockChangeLog"
	}
]
```

#### For `mongockLock` collection
```json
[
	{
		"v" : 2,
		"key" : {
			"_id" : 1
		},
		"name" : "_id_",
		"ns" : "my-database.mongockLock"
	},
	{
		"v" : 2,
		"unique" : true,
		"key" : {
			"key" : 1
		},
		"name" : "key_1",
		"ns" : "my-database.mongockLock"
	}
]
```

Try that, if one way doesn't work, let's say allowing Mongock to do it, then try creating them manually(or vice versa). And let me know how it goes

**Comment by dieppa** [comment]
The documentation has been updated. Please take a look and let me know if it makes sense to you. Thanks

**Comment by dieppa** [comment]
Hello @leonelsanchesdasilva ,did you solve it?

**Comment by leonelsanchesdasilva** [comment]
Hi @dieppa. We got to deploy the solution today, now proceeding to create the indexes. Will keep you posted after our tests. Thanks!

**Comment by dieppa** [comment]
Hello @leonelsanchesdasilva , how did it go?

**Comment by leonelsanchesdasilva** [comment]
Hi @dieppa. I'm still blocked by the offshore team. Will let you know how it was as soon as I get a response.

**Comment by rvilike** [comment]
Hello @dieppa, I'm facing the same issue, trying to make mongock with CosmosDB. I've set `index-creation=false` and created the indexes manually. The problem is that CosmosDB is sharded so it won't allow creating unique indexes without the shard key being included: [https://docs.microsoft.com/en-us/azure/cosmos-db/mongodb-indexing#unique-indexes](https://docs.microsoft.com/en-us/azure/cosmos-db/mongodb-indexing#unique-indexes)
So for example for the mongoChangeLog collection, instead of the `"executionId" : 1, "author" : 1, "changeId" : 1` I can only create `"_id": 1, "executionId" : 1, "author" : 1, "changeId" : 1` (our shard key on the collection is `"_id"`)

Unfortunatelly, it looks mongock is looking for an exact match as I receive "Index creation not allowed, but not created or wrongly created".

Is there a way I can make this work?
Thanks

**Comment by dieppa** [comment]
Hello @rvilike and @leonelsanchesdasilva ,

I am not totally clear on what it's the exact error, but doesn't seem anything unfixable. 
1. If possible, Mongock can be modified in such a way that keep working as it is with MongoDB and works fine for CosmosDB without requiring anything from the user.
2. If not, we could add some configuration to let Mongock that CosmosDB is using.
3. If that's not possible(or best), we could create another Driver for CosmosDB(Actually Drivers are made to provide a wide range of databases)


So I am confident there is a solution for it. But it requires at least a PR, may a bit more work. I cannot tell you when we can work on that, but would be very likely in November. But you're welcome to contribute and make a PR

**Comment by rvilike** [comment]
Thanks for the answer @dieppa ! After some experimenting, it turns out I can actually create unsharded collections in Cosmos, their documentation (and terraform) wasn't clear about that. Everything works fine when mongock's collections are unsharded, either I can create the right indexes or mongock also works now with index creation enabled. Moreover, we've switched to the 3.6 mongodb version in Cosmos, now mongock can create the collections as well so everything works out of the box as expected.
@leonelsanchesdasilva if you still face the issue I recommend switching to the 3.6 mongo version or use unsharded collections for mongock.

**Comment by dieppa** [comment]
That's good to hear!! But we definitely need to look at sharded collection in CosmosDB. That's a big limitation, specially in production.

Thanks!

**Comment by leonelsanchesdasilva** [comment]
@rvilike Very good hints. We will deploy again next week, and I'll verify with the offshore team how to adjust Terraform to use unsharded collections. Thanks for letting us know!

**Comment by dieppa** [comment]
I close this as I created a issue for CosmosDB

**Comment by frochi42** [comment]
Hi @dieppa Could you please add the link to the mentioned issue for CosmosDB sharded collectipns? Thanks!

**Comment by dieppa** [comment]
Hello, we still haven't done any work on this yet, but we'll start soon.
[link](https://github.com/cloudyrock/mongock/issues/262)

**Comment by frochi42** [comment]
Hi @dieppa 
just to make sure that we are talking about the same problem. 
Wenn starting my app, I get the following exception when MongoCK tries to create the indexes. Interesting that the index creation for collection mongockLock succeeded.

Thanks, Thomas

12:05:09.938 [main] DEBUG c.g.c.m.d.m.s.v.r.MongoSync4RepositoryBase:63 - Removing residual uniqueKeys for collection [mongockLock]
12:05:11.425 [main] DEBUG c.g.c.m.d.m.s.v.r.MongoSync4RepositoryBase:94 - Index in collection [mongockLock] was recreated
12:05:12.919 [main] DEBUG c.g.c.m.d.m.s.v.r.MongoSync4RepositoryBase:63 - Removing residual uniqueKeys for collection [mongockChangeLog]
12:05:13.229 [main] ERROR io.changock.runner.core.executor.ChangockBase:69 - Error in changock process. ABORTED MIGRATION
io.changock.migration.api.exception.ChangockException: com.mongodb.MongoCommandException: Command failed with error 67 (CannotCreateIndex): 'cannot create unique index over { executionId : 1.0, author : 1.0, changeId : 1.0 } with shard key pattern { changeId : 1.0 }' on server riot-pf-cosmosdb-we-dev-westeurope.mongo.cosmos.azure.com:10255. The full response is {"ok": 0.0, "errmsg": "cannot create unique index over { executionId : 1.0, author : 1.0, changeId : 1.0 } with shard key pattern { changeId : 1.0 }", "code": 67, "codeName": "CannotCreateIndex"}
	at io.changock.runner.core.executor.ChangockBase.execute(ChangockBase.java:68)
	at io.changock.runner.spring.v5.SpringApplicationRunner.run(SpringApplicationRunner.java:22)
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:786)
	at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:776)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:322)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1237)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226)
	at com.bosch.riot.automationhandler.AutomationHandlerApplication.main(AutomationHandlerApplication.java:10)
Caused by: com.mongodb.MongoCommandException: Command failed with error 67 (CannotCreateIndex): 'cannot create unique index over { executionId : 1.0, author : 1.0, changeId : 1.0 } with shard key pattern { changeId : 1.0 }' on server riot-pf-cosmosdb-we-dev-westeurope.mongo.cosmos.azure.com:10255. The full response is {"ok": 0.0, "errmsg": "cannot create unique index over { executionId : 1.0, author : 1.0, changeId : 1.0 } with shard key pattern { changeId : 1.0 }", "code": 67, "codeName": "CannotCreateIndex"}
	at com.mongodb.internal.connection.ProtocolHelper.getCommandFailureException(ProtocolHelper.java:175)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:359)
	at com.mongodb.internal.connection.InternalStreamConnection.sendAndReceive(InternalStreamConnection.java:280)
	at com.mongodb.internal.connection.UsageTrackingInternalConnection.sendAndReceive(UsageTrackingInternalConnection.java:100)
	at com.mongodb.internal.connection.DefaultConnectionPool$PooledConnection.sendAndReceive(DefaultConnectionPool.java:490)
	at com.mongodb.internal.connection.CommandProtocolImpl.execute(CommandProtocolImpl.java:71)
	at com.mongodb.internal.connection.DefaultServer$DefaultServerProtocolExecutor.execute(DefaultServer.java:255)
	at com.mongodb.internal.connection.DefaultServerConnection.executeProtocol(DefaultServerConnection.java:202)
	at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:118)
	at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:110)
	at com.mongodb.internal.operation.CommandOperationHelper.executeWriteCommand(CommandOperationHelper.java:371)
	at com.mongodb.internal.operation.CommandOperationHelper.executeWriteCommand(CommandOperationHelper.java:362)
	at com.mongodb.internal.operation.CommandOperationHelper.executeCommand(CommandOperationHelper.java:286)
	at com.mongodb.internal.operation.CommandOperationHelper.executeCommand(CommandOperationHelper.java:279)
	at com.mongodb.internal.operation.CreateIndexesOperation$1.call(CreateIndexesOperation.java:197)
	at com.mongodb.internal.operation.CreateIndexesOperation$1.call(CreateIndexesOperation.java:192)
	at com.mongodb.internal.operation.OperationHelper.withConnectionSource(OperationHelper.java:649)
	at com.mongodb.internal.operation.OperationHelper.withConnection(OperationHelper.java:611)
	at com.mongodb.internal.operation.CreateIndexesOperation.execute(CreateIndexesOperation.java:192)
	at com.mongodb.internal.operation.CreateIndexesOperation.execute(CreateIndexesOperation.java:72)
	at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.execute(MongoClientDelegate.java:195)
	at com.mongodb.client.internal.MongoCollectionImpl.executeCreateIndexes(MongoCollectionImpl.java:847)
	at com.mongodb.client.internal.MongoCollectionImpl.createIndexes(MongoCollectionImpl.java:830)
	at com.mongodb.client.internal.MongoCollectionImpl.createIndexes(MongoCollectionImpl.java:825)
	at com.mongodb.client.internal.MongoCollectionImpl.createIndex(MongoCollectionImpl.java:810)
	at com.github.cloudyrock.mongock.driver.mongodb.sync.v4.repository.MongoSync4RepositoryBase.createRequiredUniqueIndex(MongoSync4RepositoryBase.java:93)
	at com.github.cloudyrock.mongock.driver.mongodb.sync.v4.repository.MongoSync4RepositoryBase.ensureIndex(MongoSync4RepositoryBase.java:52)
	at com.github.cloudyrock.mongock.driver.mongodb.sync.v4.repository.MongoSync4RepositoryBase.initialize(MongoSync4RepositoryBase.java:37)
	at io.changock.driver.core.driver.ConnectionDriverBase.initialize(ConnectionDriverBase.java:43)
	at io.changock.runner.core.executor.MigrationExecutor.initializationAndValidation(MigrationExecutor.java:204)
	at io.changock.runner.spring.v5.core.SpringMigrationExecutor.initializationAndValidation(SpringMigrationExecutor.java:31)
	at io.changock.runner.core.executor.MigrationExecutor.executeMigration(MigrationExecutor.java:61)
	at io.changock.runner.spring.v5.core.SpringMigrationExecutor.executeMigration(SpringMigrationExecutor.java:37)
	at io.changock.runner.core.executor.ChangockBase.execute(ChangockBase.java:53)
	... 7 common frames omitted

**Comment by dieppa** [comment]
@frochi42 probably. This issue seems to be related to the fact that you're using CosmosDB, so the ticket is to address that.

**Comment by frochi42** [comment]
Hi @dieppa thanks for that confirmation. I was also pretty sure.
The problem is that I need a working solution "soon". So the question for me: Can I wait for you having corrected the problem, or do I need to explicitly create those two indexes in my SW?
Thanks a lot,
Thomas

**Comment by dieppa** [comment]
Hi again!

While we definitely have this in our agenda, I cannot commit that will be "soon". We have some other priorities first, like licensed users. So in the meanwhile, I suggest you to add the indexes yourself(look at this [documentation](https://www.mongock.io/further-configuration#index-creation))

**Comment by rvilike** [comment]
@frochi42 what wire protocol version do you use? See my answer above, 3.6 works for us - mongock works out of the box with 3.6, it can create it's collections and indexes. If you use older wire protocol version in CosmosDB, the trick is that you need to create unsharded collections for mongock

**Comment by dieppa** [comment]
@rvilike thanks a lot. We need to add this to the documentation. Any relevant comment from your experience apart from your last one?

**Comment by rvilike** [comment]
@dieppa no issues since then, things work fine!

**Comment by frochi42** [comment]
Unfortunately we cannot use unsharded collections :-(
We also are using 3.6
If you are interested: I found the reason why the index creation for the mongockChangeLog collection fails: It's the order in the Document that you pass to the createIndex() method. Obviously CosmosDB requires that the shard key comes first.
My snippet originally was:
        final Document indexDocument = new Document();
        indexDocument.append("executionId", 1);
        indexDocument.append("author", 1);
        indexDocument.append("changeId", 1);

        MongoCollection<Document> mongockChangeLogCollection = mongoDB.getCollection("mongockChangeLog");
        String result = mongockChangeLogCollection.createIndex(indexDocument, new IndexOptions().unique(true));
When doing this, I get exactly the same exception that also MongoCk gets.
If I put 
        indexDocument.append("changeId", 1);
at the beginning, then it works.

Probably, if MongoCK would change the constructor for MongoSync4ChangeEntryRepository from 
    public MongoSync4ChangeEntryRepository(MongoCollection<Document> collection, boolean indexCreation) {
         super(collection, new String[]{KEY_EXECUTION_ID, KEY_AUTHOR, KEY_CHANGE_ID}, indexCreation);
    }
to 
    public MongoSync4ChangeEntryRepository(MongoCollection<Document> collection, boolean indexCreation) {
         super(collection, new String[]{KEY_CHANGE_ID, KEY_EXECUTION_ID, KEY_AUTHOR}, indexCreation);
    }
it could work for CosmosDB. (However, I do not know whether there are side effects for other DB's)

**Comment by frochi42** [comment]
BTW: we are forced to explicitly attach the shard-key at the two MongoCK collections ("changeId" for mongockChangeLog, "key" for mongockLock). Could there be an automatic attachment by MongoCK`?
